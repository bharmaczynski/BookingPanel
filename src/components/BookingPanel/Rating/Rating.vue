<template>
    <div class="rating">
        <div class="rating__stars">
            <StarTemplate />
            <div
                v-for="index in maxRating"
                :key="index"
                class="rating__star-icon"
                :class="[starClass(index)]"
            >
                <svg width="14" height="14" viewBox="0 0 14 14">
                    <use v-if="starClass(index) === 'half'" href="#star" fill="url(#half)" />
                    <use v-else href="#star" />
                </svg>
            </div>
        </div>
        <div class="rating__number-of-rates text--bold font-size-s">
            {{ numberOfRates }}
        </div>
    </div>
</template>
<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator';
import StarTemplate from '@/components/BookingPanel/Rating/Star/starTemplate.svg';

const MAX_RAING = 5;

enum STAR_CLASSES {
    active = 'active',
    inactive = 'inactive',
    half = 'half',
}

@Component({
    components: {
        StarTemplate,
    },
})
export default class Rating extends Vue {
    maxRating: number = MAX_RAING;

    @Prop({ default: 0 })
    numberOfRates!: number;

    @Prop({ default: 0 })
    rating!: number;

    get roundHalfRating(): number {
        return Math.round(this.rating * 2) / 2;
    }

    starClass(index: number): string {
        let className = STAR_CLASSES.active;

        if (index - 0.5 === this.roundHalfRating) className = STAR_CLASSES.half;
        if (index - 0.5 > this.roundHalfRating) className = STAR_CLASSES.inactive;

        return className;
    }
}
</script>
<style lang="scss" scoped>
.rating {
    display: flex;
    align-items: flex-start;
    margin-top: 5px;

    &__number-of-rates {
        margin-left: 10px;
        margin-top: 2px;
    }

    &__stars {
        display: flex;
        align-items: flex-start;
        height: 14px;
    }

    &__star-icon {
        &.active {
            fill: $oceanGreen;
        }
        &.inactive {
            fill: $sliverSand;
        }
    }
}
</style>
