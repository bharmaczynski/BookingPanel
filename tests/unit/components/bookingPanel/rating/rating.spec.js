import { shallowMount } from '@vue/test-utils';
import Rating from '@/components/bookingPanel/rating/Rating.vue';

describe('Rating', () => {
    let propsData = {
        numberOfRates: 40,
        rating: 3.7,
    };

    function createWrapper(propsData = {}) {
        return shallowMount(Rating, { propsData });
    }

    describe('Snapshots', () => {
        it('should match initial snapshot', () => {
            const wrapper = createWrapper();

            expect(wrapper.vm.$el).toMatchSnapshot();
        });

        it('should match snapshot with props', () => {
            const wrapper = createWrapper(propsData);

            expect(wrapper.vm.$el).toMatchSnapshot();
        });
    });

    describe('Methods', () => {
        describe('starClass', () => {
            it('should return "active" for number less than rounded rating', () => {
                const wrapper = createWrapper(propsData);

                expect(wrapper.vm.starClass(3)).toEqual('active');
            });

            it('should return "half" for number which after subtraction 0.5 is equal rounded rating ', () => {
                const wrapper = createWrapper(propsData);

                expect(wrapper.vm.starClass(4)).toEqual('half');
            });

            it('should return "inactive" for number which after subtraction 0.5 is greater rounded rating ', () => {
                const wrapper = createWrapper(propsData);

                expect(wrapper.vm.starClass(5)).toEqual('inactive');
            });
        });
    });
});
