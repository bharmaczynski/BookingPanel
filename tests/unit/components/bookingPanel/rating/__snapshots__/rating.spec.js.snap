// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating Snapshots should match initial snapshot 1`] = `
<div
  class="rating"
>
  <div
    class="rating__stars"
  >
    <!---->
     
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
  </div>
   
  <div
    class="rating__number-of-rates text--bold font-size-s"
  >
    
        0
    
  </div>
</div>
`;

exports[`Rating Snapshots should match snapshot with props 1`] = `
<div
  class="rating"
>
  <div
    class="rating__stars"
  >
    <!---->
     
    <div
      class="rating__star-icon active"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon active"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon active"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon half"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          fill="url(#half)"
          href="#star"
        />
      </svg>
    </div>
    <div
      class="rating__star-icon inactive"
    >
      <svg
        height="14"
        viewBox="0 0 14 14"
        width="14"
      >
        <use
          href="#star"
        />
      </svg>
    </div>
  </div>
   
  <div
    class="rating__number-of-rates text--bold font-size-s"
  >
    
        40
    
  </div>
</div>
`;
